qemu/tests/chardev_free_port.py

```diff
@@ -0,0 +1,51 @@
+from virttest import error_context, env_process
+from virttest import utils_misc
+
+from qemu.tests.virtio_console import add_chardev
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    qemu should try to find a free port value by to=<port> with unix socket and tcp options:
+    1) boot guest with socket 'host=127.0.0.1,port=1234'
+    2) query chardev and check port number
+    3) boot another guest with socket 'host=127.0.0.1,port=1234,to=2234'
+    4) query chardev and check port number, should different from 2)
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    chardev_infos = []
+    vms = params.get("vms").split()
+    _vm = env.get_vm(vms[0])
+    char_device = add_chardev(_vm, params)[0]
+    chardev_params = char_device.params
+    for vm_ind, vm in enumerate(vms):
+        if vm_ind == 1:
+            host = chardev_params['host']
+            chardev_to = utils_misc.find_free_ports(int(chardev_params['port']) + 1, 6000, 1, host)
+            chardev_params['to'] = str(chardev_to[0])
+
+        extra_params = ' ' + char_device.cmdline()
+        params['extra_params_%s' % vm] = params.get('extra_params', '') + extra_params
+        params['start_vm_%s' % vm] = "yes"
+        env_process.preprocess(test, params, env)
+        _vm = env.get_vm(vm)
+        chardev_infos.append(_vm.monitor.info("chardev"))
+
+    for char_ind, chardevs in enumerate(chardev_infos):
+        in_chardev = False
+        for chardev in chardevs:
+            if chardev['label'] == chardev_params['id']:
+                tmp_pnum = int(chardev['filename'].split(':')[-1].split(',')[0])
+                break
+        if char_ind == 0:
+            if tmp_pnum == int(chardev_params['port']):
+                in_chardev = True
+        else:
```

Comment:
error_context.context("The expect port for vm%s is in (%s, %s]" %(char_ind, port_in_param, to_in_pram), logging.info)